eNMS/controller/base.py

```diff
@@ -692,3 +693,8 @@ def update_database_configurations_from_git(self):
                 for property in self.configuration_properties
             ):
                 pool.compute_pool()
+
+
+@atexit.register
+def cleanup():
+    db.engine.dispose()  # gracefully close database connections

```

Comment:
can you move that inside the database class ? (using `register` in `__init__` because decorator won't work for class function I guess)

Reply:
Yes, sir.  I'll check this in a minute - that makes much more sense.

The decorator does support using a @staticmethod (i.e., function that does not require arguments), but the method implementation needs access to the `db` module variable. 

There is an alternative - with an extra static class variable.  If you prefer this, let me know and I'll do this instead:

```python
class Database:
    db = None
    
    @staticmethod
    @atexit.register
    def cleanup():
        Database.db.engine.dispose()  # gracefully close database connections

db = Database.db = Database() 
```

Reply:
I'm sure it would work too, but staticmethod suggests that the function doesn't use `self`, when in fact it does as we are calling `self.engine.dipose()`. It's somewhat inelegant to use a staticmethod that uses `self` via an alternate variable that has the same value as `self`.
So ideally we would use `@atexit.register` without `staticmethod` and pass `self` as first argument, but I'm afraid that will fail

Reply:
So I wrote a little Python test to make sure - an instance method  with @atexit.register produces `TypeError: cleanup() missing 1 required positional argument: 'self'`

I'm looking for an alternative on stackoverflow to see if maybe there is an existing solution.

Reply:
So, if I call the atexit() more directly, I can supply arguments.

Is this more appealing?

```python
import atexit

class Database:
    def __init__(self):
        # regular initialization
        atexit.register(self.cleanup, self)

    @staticmethod
    def cleanup(db):  # db passed as argument
        db.engine.dispose()
```

Reply:
just register(self.cleanup) and remove the staticmethod, then self.engine.dispose()

Reply:
Of course- that makes sense - checking that in here in a second!  
Thanks Antoine!

Reply:
Made a mistake, forgot to change db.engine back to self.engine - repushed just now.